library/core/src/intrinsics.rs

```diff
@@ -1840,20 +1840,23 @@ pub(crate) fn is_nonoverlapping<T>(src: *const T, dst: *const T, count: usize) -
 /// [`Vec::append`]: ../../std/vec/struct.Vec.html#method.append
 #[doc(alias = "memcpy")]
 #[stable(feature = "rust1", since = "1.0.0")]
+#[rustc_const_unstable(feature = "const_intrinsic_copy", issue = "none")]
 #[inline]
-pub unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize) {
+pub const unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize) {
     extern "rust-intrinsic" {
+        #[rustc_const_unstable(feature = "const_intrinsic_copy", issue = "none")]
         fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize);
     }
 
-    if cfg!(debug_assertions)
+    // FIXME: Is it possible to make this work in const fn?
+    /*if cfg!(debug_assertions)
         && !(is_aligned_and_not_null(src)
             && is_aligned_and_not_null(dst)
             && is_nonoverlapping(src, dst, count))
```

Comment:
Is it possible to make this work in const fn? Is there any overlap with what is done in compiler/rustc_mir/src/interpret/intrinsics.rs, or is that only affecting CTFE?

Reply:
`is_aligned_and_not_null()` and `is_nonoverlapping()` both seems to do some `*const _ as usize` plus math which I believe is a no no during CTFE?

Reply:
For `is_nonoverlapping` you could create another intrinsic that implements the logic in the const evaluator, but then you'd have to implement a normal codegen version of that, too.

The other two can actually be implement in a const-evaluable way with https://doc.rust-lang.org/std/primitive.pointer.html#method.align_offset

That said... Since the const-eval intrinsic checks all these invariants *anyway*, what we can do is to finally introduce a way to run different code at runtime vs at compile-time https://github.com/rust-lang/const-eval/issues/7#issuecomment-576243335

This is as good as a PR to do that as any. Do you want to tackle that or is it more than you wanted on your plate right now? It may be more effective to do just the new intrinsic in a separate PR so we have some unit tests before we use it in `copy_nonoverlapping`

Reply:
I am *really* new working on the compiler and do not really know what to do. If you think it might be a not-too-great first thing to implement then I will happily start with something else. Otherwise, if I knew where to start and given some (probably lots of) pointers along the way... Maybe I could give it a try, but I would not want to promise anything and am not sure about when and how much time I can spend on it. I certainly would not want to stand in the way for anyone else who want to give it a go.

Reply:
I just did a dive into the impl, and while the const eval parts are fairly self contained, the codegen parts are not, so... I don't think we can make `copy_nonoverlapping` `const fn` for now, but `copy` could still be doable by replacing the implementation at https://github.com/rust-lang/rust/blob/d015f0d92144f0e72735a918aee8510b0fe2cff5/library/core/src/intrinsics.rs#L1749 with a `const fn` capable one. This should be doable by using `guaranteed_ne` for the null pointer check and `align_offset` for the alignment check. (you should be able to experiment with this on the playground, so you don't have to keep recompiling libcore).

Reply:
Ok thanks, I will take a look :)

Reply:
Did you mean `ptr::align_offset`? It looks to me like that function contains a whole lot of math on `ptr as usize` or am I missing something? :)

https://github.com/rust-lang/rust/blob/d015f0d92144f0e72735a918aee8510b0fe2cff5/library/core/src/ptr/mod.rs#L1147

Reply:
Oh wait there is

https://github.com/rust-lang/rust/blob/d015f0d92144f0e72735a918aee8510b0fe2cff5/library/core/src/ptr/const_ptr.rs#L851

What is (or where can I read about) the idea with `library/core/src/ptr/mod.rs` vs `library/core/src/ptr/const_ptr.rs`. Is the `const_ptr.rs` magically picked during CTFE?

Reply:
@oli-obk 
> This should be doable by using guaranteed_ne for the null pointer check and align_offset for the alignment check.

Did you mean `ptr::align_offset` or did you have something else in mind?

Reply:
I did mean that... and i must apologize... I completely mixed up a bunch of things.

`const_ptr.rs` and `mut_ptr.rs` are `*const T` and `*mut T` respectively. They have nothing to do with const eval, except that some of their functions are `const fn`.

`align_offset` was a big topic around CTFE some time ago, but we only ever ended up landing it in miri: https://github.com/rust-lang/miri/blob/510e62c9ea5c9da6c7b29ba89568456e035f70c6/src/shims/mod.rs#L36

So... before you can do this PR, we'd need to move the `align_offset` logic from miri to rustc and thus CTFE. This is not as easy as with other low level intrinsics, because `align_offset` is a `lang_item` which contains the runtime code, but different code from the function body will be run in the CTFE/miri case. While miri has this working already, and it would be simple to use the same trick as the one linked above, it would still not compile, because our checks would see non-const code in the `const fn align_offset`. So we would need to create a new system which makes a function `const` from the outside, but not `const` on the inside. I'm not sure about the extend of the work necessary to do that, but it's probably as simple as not const-checking the function body of lang items.

Reply:
No problem! :)

Thanks for the insight into `{const, mut}_ptr.rs`.

Would it make most sense to close this PR until `align_offset` has been constified then?

Reply:
I guess so. I see no way to implement this now.

Reply:
Ok, thank you :)

Reply:
@oli-obk
> So we would need to create a new system which makes a function const from the outside, but not const on the inside. I'm not sure about the extend of the work necessary to do that, but it's probably as simple as not const-checking the function body of lang items.

Would something like the following do?

 [compiler/rustc_mir/src/transform/check_consts/validation.rs:322](https://github.com/rust-lang/rust/blob/master/compiler/rustc_mir/src/transform/check_consts/validation.rs#L322)
```diff
 if self.tcx.sess.opts.debugging_opts.unleash_the_miri_inside_of_you {
     self.tcx.sess.miri_unleashed_feature(span, gate);
     return;
 }
+
+let attributes = self.tcx.get_attrs(self.def_id().to_def_id());
+if self.tcx.sess.contains_name(&attributes, sym::lang) {
+    // Disable const check for lang items
+    // FIXME: What should be done here?
+    return;
+}

let mut err = op.build_error(self.ccx, span);
assert!(err.is_error())
```

Seems like a very powerful change (if it even works) with risk of accidental usage of non-const things? Might it make more sense to have a new attribute `#[unleash_the_miri_inside_of_you_here]` to explicitly mark where this behavior is wanted?

Just speculating, I am quite far outside my comfort zone, and don't understand any far reaching implications :)

Reply:
> Might it make more sense to have a new attribute #[unleash_the_miri_inside_of_you_here] to explicitly mark where this behavior is wanted?

:laughing: I love it. I would definitely accept a PR that adds this (forever and ever unstable) attribute, even if we don't end up using it for `align_offset`. As an initial step I would use it for the `ptr_guaranteed_eq` and `ptr_guaranteed_ne` intrinsics by making them lang items instead. This way we can get rid of the `codegen_ssa` implementation of these intrinsics, because runtime will just use the function body, while CTFE will just never load the body but run its own logic instead.

And yes, you found the right place to do this.

Reply:
Would it even be possible to implement the proposed
```rust
pub fn const_eval_select<ARG, F, G, RET>(arg: ARG, called_in_const: F, called_at_rt: G) -> RET;
```
as an "unleashed" `const fn`?

Reply:
good idea, that could be doable, but we'd need a `Fn` bound on `G`, not sure if that triggers some other checks.

Reply:
I tested with this but i still get `calling non-const function `non_const_fn`, however it does seem to notice the attribute.

```rust
let attributes = self.tcx.get_attrs(self.def_id().to_def_id());
if self.tcx.sess.contains_name(&attributes, sym::unleash_the_miri_inside_of_you_here) {
    // FIXME: What should be done here?
    println!("Yay I noticed the unleash_the_miri_inside_of_you_here attribute!!!");
    return;
}
```

Is there some code to add before the return to tell the compiler that everything is fine or is the return what tells it that all *is* fine?

<details>
  <summary>Test Code</summary>

  ```rust
  #![feature(rustc_attrs)]
  
  fn main() {
      const FOO: bool = call_non_const_fn_unleashed();
      assert!(FOO);
  }
 
  #[unleash_the_miri_inside_of_you_here]
  const fn call_non_const_fn_unleashed() -> bool {
      non_const_fn()
  }
  
  fn non_const_fn() -> bool { true }

  ```
</details>

<details>
  <summary>Error from ./x.py test src/test/ui/unleash-the-miri-inside-of-you-here</summary>

  ```
  ---- [ui] ui/unleash-the-miri-inside-of-you-here/call-non-const-from-const-fn.rs stdout ----
  normalized stdout:
  Yay I noticed the unleash_the_miri_inside_of_you_here attribute!!!
  
  
  
  The actual stdout differed from the expected stdout.
  Actual stdout saved to /home/albin/rust/rustc/build/x86_64-unknown-linux-gnu/test/ui/unleash-the-miri-inside-of-you-here/call-non-const-from-const-fn/call-non-const-from-const-fn.stdout
  normalized stderr:
  error: any use of this value will cause an error
    --> $DIR/call-non-const-from-const-fn.rs:12:5
     |
  LL |     const FOO: bool = call_non_const_fn_unleashed();
     |     ------------------------------------------------
  ...
  LL |     non_const_fn()
     |     ^^^^^^^^^^^^^^
     |     |
     |     calling non-const function `non_const_fn`
     |     inside `call_non_const_fn_unleashed` at $DIR/call-non-const-from-const-fn.rs:12:5
     |     inside `FOO` at $DIR/call-non-const-from-const-fn.rs:5:23
  ```
</details>

Reply:
We've reached a point in the design where I think we need to invoke @rust-lang/wg-const-eval and see what they think of it.

> Is there some code to add before the return to tell the compiler that everything is fine or is the instant return what tells it that all is fine?

No, because *it is already working*! You did everything right, you can notice this because `FOO` actually is getting evaluated now, when before it errored before evaluation. The error you are seeing is an actual evaluation error with a backtrace and everything. We have a second safety net for catching calls to non-const fn during evaluation, and we don't want to change this. What we do want is to catch the call to `call_con_const_fn_unleashed` and not even evaluate its body, but run special code within the const evaluator.

Initially I thought we'd have to make such functions a lang item *in addition* to `unleash_the_miri_inside_of_you_here`, but I think I have an alternative idea that could work: we change the attribute to also take a string argument, so we have `unleash_the_miri_inside_of_you_here(guaranteed_eq)` and then fetch that attribute in https://github.com/rust-lang/rust/blob/3d9ada686fb42bd036b3a4916526f413f1d5d1f8/compiler/rustc_mir/src/const_eval/machine.rs#L213 and if it's one of a list of things that we know (match on the symbol, you'll have to register the symbol in https://github.com/rust-lang/rust/blob/e4297ba39ce07a0b0000841b50154ea53783a024/compiler/rustc_span/src/symbol.rs#L123 ), we do the special logic, and otherwise we go back and continue with the current https://github.com/rust-lang/rust/blob/3d9ada686fb42bd036b3a4916526f413f1d5d1f8/compiler/rustc_mir/src/const_eval/machine.rs#L217 (which we'll be able to replace with your new scheme, too, which is great imo).

Reply:
Ok, I'll try to wrap my head around that :)

But essentially you are saying that my changes this far removes the check *before* the actual const eval. So with the attribute on, the actual const eval gets to decide?

Would that mean that const eval would be happy with (ignoring the bounds and stuff for now)
```rust
#[unleash_the_miri_inside_of_you_here]
fn const_eval_select<ARG, F, G, RET>(arg: ARG, called_in_const: F, called_at_rt: G) -> RET {
    if this_is_run_in_const_eval() {
        called_in_const(arg)
    } else {
        called_at_rt(arg)
    }
}
```
since it never actually tries to call `called_at_rt`?

---

> What we do want is to catch the call to call_con_const_fn_unleashed and not even evaluate its body, but run special code within the const evaluator.

Wait, so we are not just catching the call to `non_const_fn`, but the entire call to `call_non_const_fn_unleashed`? So then the `const_eval_select` thing would not work?

Reply:
> Wait, so we are not just catching the call to `non_const_fn`, but the entire call to `call_non_const_fn_unleashed`?

Yes. Otherwise we'd need *another* system to catch calls to `non_const_fn`, and this way we can just catch the `call_non_const_fn_unleashed`. If you need some preprocessing, you can do that ahead of time in a regular const fn that calls `call_non_const_fn_unleashed`.

> So then the `const_eval_select` thing would not work?

I think you would write it as

```rust
#[unleash_the_miri_inside_of_you_here(const_eval_select)]
fn const_eval_select<ARG, F, G, RET>(arg: ARG, called_in_const: F, called_at_rt: G) -> RET {
    called_at_rt(arg)
}
```

and then the CTFE engine has to process that function and can't just keep evaluating on.

But you are right, we could also do this without catching, if we made `this_is_run_in_const_eval` a `unleash_the_miri_inside_of_you_here` and caught only that and returned `true` from CTFE but the function is declare as

```rust
#[unleash_the_miri_inside_of_you_here]
fn this_is_run_in_const_eval() -> bool {
    false
}
```

I guess we could do it this way, but it will require some care because we have never done function calls via generics in const eval before, I'm not sure what problems you would encounter. We should carefully try out all of these different use cases in individual PRs that make it easy to look at all effects of just the single change

Reply:
> We've reached a point in the design where I think we need to invoke @rust-lang/wg-const-eval and see what they think of it.

Is there a summary of the current proposal?

My first reaction upon skimming the discussion here is: I feel like this is RFC material. There's clearly a large design space, and having the design documented seems important. We don't want to accidentally subvert our CTFE safety net.

"unleash-miri" was meant as a never-stable hack that lets us test the safety net, I don't think it should become an attribute ever.

Sorry to be the bearer of bad news here, but I don't think having such important design discussions in a thread in a PR whose title is about something entirely different is a good way to go about this.

Reply:
> Sorry to be the bearer of bad news here, but I don't think having such important design discussions in a thread in a PR whose title is about something entirely different is a good way to go about this.

I fully agree :) This PR is probably far from a good place for the discussion. However I do not think I am anywhere near qualified to make an RFC on the subject. However I am here should there be anything I *can* help with :)

Reply:
Would it be useful to continue working on this PR draft, or open a more restricted one, as an example use case/background for the problem and one potential solution?

Reply:
For this PR, I see two options:
- Leave it as "something we can do once we have a story for const-dependent dispatch".
- Comment out the debug assertions for now. Their usefulness is anyway limited since the libstd everyone uses is compiled without debug assertions.

I guess the question is one of evaluating the relative usefulness of these new const operations vs the assertions.

Reply:
> For this PR, I see two options:
> 
> * Leave it as "something we can do once we have a story for const-dependent dispatch".
> * Comment out the debug assertions for now. Their usefulness is anyway limited since the libstd everyone uses is compiled without debug assertions.
> 
> I guess the question is one of evaluating the relative usefulness of these new const operations vs the assertions.

I have constified `ptr::read[_unaligned]` and `ptr::write` locally by commenting out the call to `is_nonoverlapping` and `align_offset`. I assume we can always remove the comments to bring back the checks and wait for const-dependent dispatch should we go for the latter option, right? `ptr::read[_unaligned]` requires #79621, should I rebase and push my changes (does not include the `unleash_the_miri_inside_of_you_here` attribute)?

Reply:
> ptr::read[_unaligned] requires #79621, should I rebase and push my changes (does not include the unleash_the_miri_inside_of_you_here attribute)?

yes, that sounds right

Reply:
> by commenting out the call to is_nonoverlapping and align_offset

Do you mean `is_aligned_and_not_null`? Which `align_offset` call?